### DISCUSSION TITLE
How to close CMD box, but keep DNSCrypt-Proxy running without installing it as a service?

### DISCUSSION THREAD

**Discussion by HqT65W6R** [DNSCrypt/dnscrypt-proxy#2737] [Q&A]
I want to just start DNSCrypt-Proxy and keep it running without installing it as a service, but I don't want CMD box to continue running in the background on desktop or even minimized. I want it closed, out-of-sight, but not close DNSCrypt-Proxy process.  How can I do that? So far only installating DNSCrypt-Proxy as service allows for that...

**Comment by jedisct1** [comment]
### How to Run a Command-Line Tool in Windows and Keep It Running in the Background

To run a command-line tool in Windows and keep it running in the background after closing the terminal, you can use one of the following methods:

---

#### **Method 1: Using `start` Command**
1. Open Command Prompt.
2. Use the `start` command to launch the program:
   ```cmd
   start "" /b path\to\your\program.exe
   ```
   - The `""` provides a placeholder for the window title (required).
   - The `/b` flag starts the program in the background without creating a new window.

   **Example:**
   ```cmd
   start "" /b notepad.exe
   ```

   After running the command, you can safely close the terminal.

---

#### **Method 2: Using `cmd` Detachment**
1. Run the following command:
   ```cmd
   cmd /c start /b path\to\your\program.exe
   ```
   - The `/c` flag tells `cmd` to execute the command and then terminate.

   **Example:**
   ```cmd
   cmd /c start /b notepad.exe
   ```

---

#### **Method 3: Running as a Background Process Using PowerShell**
If you're using PowerShell, you can run the following command:
```powershell
Start-Process -NoNewWindow -FilePath "path\to\your\program.exe"
```

**Example:**
```powershell
Start-Process -NoNewWindow -FilePath "notepad.exe"
```

This starts the process without creating a new window.

---

#### **Method 4: Creating a Scheduled Task (Advanced Option)**
If the program needs to run persistently in the background (e.g., surviving reboots):
1. Open **Task Scheduler**.
2. Create a new task.
3. Set the trigger to **"At startup"** or **"On demand"**.
4. Set the action to **"Start a program"** and provide the path to the program.
5. Save the task and run it.

---

These methods will ensure that your command-line tool continues running even after the terminal is closed.

**Comment by HqT65W6R** [comment]
Have you tried these methods with DNSCrypt-Proxy.exe yourself? All your methods start it, but DNSCrypt-Proxy.exe terminates itself as soon as CLI is closed, regardless of whether its CMD or PS.

**Comment by jedisct1** [comment]
Try the Task Scheduler method, then:

Open Task Scheduler.
Create a new task:
Trigger: "At startup" or manually as needed.
Action: "Start a program" -> Provide the full path to dnscrypt-proxy.exe.
Ensure "Run whether user is logged on or not" is selected.
Save the task and run it.

This method makes dnscrypt-proxy independent of the terminal.

**Comment by lifenjoiner** [comment]
I answered a similar question before: https://github.com/lifenjoiner/pd/issues/9#issuecomment-2324445330 .

1. [A software that puts any application to the system tray](https://www.reddit.com/r/software/comments/mskgvn/a_software_that_puts_any_application_to_the/)

    Recommendations:
    * https://github.com/plastovicka/HotkeyP
    * https://github.com/fcFn/traymond

2. totally hidden by 3rd party Apps or scripts
    Save the following code as `dnscrypt-proxy.bat` beside `dnscrypt-proxy.exe`, and then run it.
```bat
@echo off
setlocal EnableDelayedExpansion
set argv=%*
if "%~1"=="" (
    set argv=%~dpn0.exe
)
set argv=!argv:\=\\\\!
set argv=!argv: =%%20!
mshta javascript:new%%20ActiveXObject('Wscript\.Shell').Run('!argv!',0);close()
```

**Reply by jedisct1** [reply]
Maybe it would be worth adding to the wiki?

**Reply by lifenjoiner** [reply]
Added https://github.com/DNSCrypt/dnscrypt-proxy/wiki/Installation-Windows#optional-run-the-proxy-as-a-command-line-program-instead-of-a-system-service

**Comment by HqT65W6R** [comment]
That's great, but is there some way to do it without relying on other tools? MSHTA is an obsolete component frequently removed for safety reasons.

**Reply by lifenjoiner** [reply]
That is "因噎废食" in Chinese idiom.

**Reply by flawrist** [reply]
Save it, for example, as `dnscrypt-start.vbs`, put into DNSCrypt folder and execute.

``` vbscript
REM dnscrypt-start.vbs
REM It silently launches dnscrypt-proxy.exe without any windows

ScriptDir = CreateObject("Scripting.FileSystemObject").GetParentFolderName(WScript.ScriptFullName)
Set WshShell = CreateObject("WScript.Shell") 
WshShell.Run Chr(34) & ScriptDir & "\dnscrypt-proxy.exe" & Chr(34), 0
Set WshShell = Nothing
```
It is expected to work on any Windows released in the 21st century. On Windows 11, it will work [until 2027](https://superuser.com/questions/1851694/what-does-dism-online-add-capability-capabilitynamevbscript-do-to-instal), when Microsoft is going to to remove VBScript. But a magical command `DISM /Online /Add-Capability /CapabilityName:VBSCRIPT~~~~` will bring the party back.

**Comment by lifenjoiner** [comment]
My first solution is multi-functional.
Now, 2 hacker ways, and mono-functional:
1. Compile the program using `go build -ldflags="all=-s -w -H windowsgui" -trimpath`.
2. Change the `Subsystem` flag of the PE optional header from "Windows Console" to "Windows GUI" using a PE tool.
